benchmark/scripts/report.ts

```diff
@@ -0,0 +1,314 @@
+          size: 20,
+        },
+      },
+      legend: {
+        display: false,
+      },
+    },
+  };
+
+  const data: ChartConfiguration["data"] = {
+    labels: props.labels,
+    datasets: [
+      {
+        label: "offset",
+        data: props.offsetData,
+        backgroundColor: "rgba(255, 255, 255, 0)",
+        borderSkipped: true,
+        datalabels: {
+          labels: {
+            title: null,
+          },
+        },
+      },
+      {
+        label: "left",
+        data: props.leftData,
+        backgroundColor: props.backgroundColors,
+        borderColor: props.borderColors,
+        minBarLength: 1.5,
+        borderWidth: {
+          right: 1.5,
+        },
+        datalabels: {
+          align: "center",
+          anchor: "end",
+          clamp: true,
+          formatter: (_, context: Context) =>
+            props.datalabels[context.dataIndex],
+        },
+      },
+      {
+        label: "right",
+        data: props.rightData,
+        backgroundColor: props.backgroundColors,
+        datalabels: {
+          labels: {
+            title: null,
+          },
+        },
+      },
+    ],
+  };
+
+  // Create a new Chart.js instance
+  const width = 800;
+  const height = 600;
+  const backgroundColour = "white";
+  const chartCallback: ChartCallback = (ChartJS) => {
+    ChartJS.defaults.color = "rgba(31, 35, 40)";
+    ChartJS.defaults.font.size = 12;
+    ChartJS.defaults.responsive = false;
+    ChartJS.defaults.maintainAspectRatio = false;
+  };
+  const chartJSNodeCanvas = new ChartJSNodeCanvas({
+    width,
+    height,
+    backgroundColour,
+    plugins: {
+      modern: ["chartjs-plugin-datalabels"],
+    },
+    chartCallback,
+  });
+  const buffer = await chartJSNodeCanvas.renderToBuffer({
+    type: "bar",
+    data,
+    options,
+  });
+  await fsPromises.writeFile(filePath, buffer, "base64");
+};
+
+function generateCharts(
+  datasetType: string,
+  datasets: Record<string, Record<string, any>>,
+  gitBranch: string
+) {
+  for (const key in datasets) {
+    const labels: string[] = [];
+    const datalabels: string[] = [];
+    const backgroundColors: string[] = [];
+    const borderColors: string[] = [];
+    const offsetData: number[] = [];
+    const leftData: number[] = [];
+    const rightData: number[] = [];
+    let max = 0;
+    let min = Infinity;
+    console.log(`  - ${datasetType} : ${key}`);
+    const dataset = Object.fromEntries(
+      Object.entries(datasets[key as keyof typeof datasets]).sort(
+        (a, b) => a[1].order - b[1].order
+      )
+    );
+    for (const key in dataset) {
+      if (dataset[key].data) {
+        const { label, mean, standardDeviation } = dataset[key];
+        labels.push(label);
+        datalabels.push(`${mean}±${standardDeviation}`);
+        backgroundColors.push(
+          CONFIG.functions.reduce(
+            (prev, curr) =>
+              curr.functionName === key ? curr.chart.backgroundColor : prev,
+            ""
+          )
```

Comment:
I'm confused - why are we reducing to determine this? Taking the last?

Reply:
You are right. I got carried away on that. A simple loop will do !